librepo/gpg_rpm.c

```diff
@@ -0,0 +1,613 @@
+write_memory_to_file_fd(int fd, const guint8 * buf, size_t len, GError **err) {
+    const ssize_t ret = write(fd, buf, len);
+    if (ret == -1) {
+        const gchar * errstr = g_strerror(errno);
+        g_debug("%s: Error writing to file descriptor %i: %s", __func__, fd, errstr);
+        g_set_error(err, LR_GPG_ERROR, LRE_IO, "Error writing to file descriptor %i: %s", fd, errstr);
+        return -1;
+    }
+    if (ret != len) {
+        g_debug("%s: Requested to write %li octets, but written %li.", __func__, (long)len, (long)ret);
+        g_set_error(err, LR_GPG_ERROR, LRE_IO, "Requested to write %li octets, but written %li", (long)len, (long)ret);
+        return -1;
+    }
+    return ret;
+}
+
+static ssize_t
+write_memory_to_file(const char * path, const guint8 * buf, size_t len, GError **err) {
+    const int fd = creat(path, S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);
+    if (fd == -1) {
+        const gchar * errstr = g_strerror(errno);
+        g_debug("%s: Error: Failed to create file \"%s\": %s", __func__, path, errstr);
+        g_set_error(err, LR_GPG_ERROR, LRE_IO, "Failed to create file \"%s\": %s", path, errstr);
+        return -1;
+    }
+    const ssize_t ret = write_memory_to_file_fd(fd, buf, len, err);
+    close(fd);
+    return ret;
+}
+
+// Searches for a key with `keyid` in the OpenPGP packet.
+static gboolean
+search_key_id(const guint8 * keyid, gchar * buf, size_t len, pgpDigParams * dig_params, GError **err) {
+    pgpDigParams main_dig_params = NULL;
+    if (pgpPrtParams((const uint8_t *)buf, len, PGPTAG_PUBLIC_KEY, &main_dig_params) == -1) {
+        g_debug("%s: Error: Parsing a OpenPGP packet(s) failed", __func__);
+        g_set_error(err, LR_GPG_ERROR, LRE_GPGERROR, "Parsing a OpenPGP packet(s) failed");
+        return FALSE;
+    }
+    if (main_dig_params == NULL) {
+        g_warning("%s: No key found in the key file.", __func__);
+        return FALSE;
+    }
+
+    gboolean ret = FALSE;
+    int subkeys_count = 0;
+    pgpDigParams * subkeys = NULL;
+    do {
+        if (memcmp(pgpDigParamsSignID(main_dig_params), keyid, sizeof(pgpKeyID_t)) == 0) {
+            if (dig_params != NULL) {
+                *dig_params = g_steal_pointer(&main_dig_params);
+            }
+            ret = TRUE;
+            break; // key id match main_dig_params
+        }
+
+        if (pgpPrtParamsSubkeys((const uint8_t *)buf, len, main_dig_params, &subkeys, &subkeys_count) == -1) {
+            g_warning("%s: Parse subkey parameters from OpenPGP packet(s) failed", __func__);
+            break;
+        }
+
+        for (int idx = 0; idx < subkeys_count; ++idx) {
+            if (memcmp(pgpDigParamsSignID(subkeys[idx]), keyid, sizeof(pgpKeyID_t)) == 0) {
+                if (dig_params != NULL) {
+                    *dig_params = g_steal_pointer(&subkeys[idx]);
+                }
+                ret = TRUE; // key id match subkey
+                break;
+            }
+        }
+    } while (FALSE);
```

Comment:
This seems just weird and confusing to me. I understand this has its meaning in macros and such, but why don't just refactor the code here into separate static function?

Reply:
The single-pass loop here allows `break` to be used instead of `goto`.
It could be solved in another way. For example, with more `if`s. But that would be rather more complicated and harder to read. Or replace `break` with `return`. But then I would have to ensure that the resources are freed before each `return`.

Reply:
Thanks.

Reply:
OK, I am not that familiar with conventions in C, so if it's commonly used in such cases, let's keep it as it is.